Use new format of MAGIC_NUMBER

eugulixes opened the pull request:
The format allows storing a character with all its attributes. However, the current implementation is restricted by storing a character and 8 colors for a background and text so far.

eugulixes committed:
Use new format of MAGIC_NUMBER

The format allows storing a character with all its attributes. However, the current implementation is restricted by storing a character and 8 colors for a background and text so far.

eugulixes committed:
Adopt tests to new format of MAGIC_NUMBER

eugulixes committed:
Wrap code related to setting bg and fg colors into methods

eugulixes committed:
Switch color scheme only when setting fg color

eugulixes committed:
Implement bold (extra bright) mode

Use Source Code Pro by default.

eugulixes committed:
Implement underline mode

eugulixes committed:
Implement blink mode

eugulixes committed:
Make cosmetic changes in dumphtml

* Amend comments
* Rename `h` and `w` into `rows` and `cols`

eugulixes committed:
Use corresponding methods for setting and cleaning bits

dshil commented on gits/terminal.py:
```diff
@@ -204,6 +235,14 @@ def _ignore(self):
+        """Sets the specified `_sgr` bit. """
+        self._sgr |= 1 << bit
+
+    def _claen_bit(self, bit):
```
s/claen/clean/

dshil commented on gits/terminal.py:
```diff
@@ -213,46 +252,66 @@ def _default_rendition(self):
         self._cap_set_color(0)
 
+    def _set_bg_color(self, color):
+        """Sets background color. """
```
s/Sets background color/Sets the background color?

dshil commented on gits/terminal.py:
```diff
@@ -213,46 +252,66 @@ def _default_rendition(self):
+        color_bits, _ = divmod(self._sgr, MAGIC_NUMBER)
+        _, fg = divmod(color_bits, 16)
+        new_color_bits = color * 16 + fg
+        self._sgr &= ~(color_bits << 40)  # clear color bits
```
Could we replace this bits manipulation with something similar to `_set_bit` or `_clean_bit`. 
E.g.
```python
def _set_bits(self, src, bit):
    self._sgr |= src << bit
```
Usage:
```python
self._set_bits(color_bits, 40)
```

dshil commented on gits/terminal.py:
```diff
@@ -213,46 +252,66 @@ def _default_rendition(self):
             pass
-        elif colour == 5:  # blink
+        elif color == 4:  # smul, start underscore mode
+            self._set_bit(UNDERLINE_BIT)
```
s/UNDERLINE_BIT/UNDERSCORE_BIT/?

Reason:
- for bold we have `BOLD_BIT`
- for blink we have `BLINK_BIT`

I just want to use the unify format for naming these constants.

eugulixes replied:
Different sources related to terminals use `underline` or `underscore`. For example, `man infocmp` uses `underline` and [1] uses both. I suggest we rename `underscore` into `underline` because the latter is well-spread in our code.

[1] http://pubs.opengroup.org/onlinepubs/7908799/xcurses/terminfo.html

dshil commented on gits/terminal.py:
```diff
@@ -213,46 +252,66 @@ def _default_rendition(self):
             pass
-        elif colour == 24:  # rmul
+        elif color == 24:  # rmul, end underscore mode
+            self._claen_bit(UNDERLINE_BIT)
```
s/claen/clean/

dshil commented on gits/terminal.py:
```diff
@@ -213,46 +252,66 @@ def _default_rendition(self):
+        elif color == 39:
+            self._sgr = BLACK_AND_WHITE
+        elif 40 <= color <= 47:  # setab
+            self._set_bg_color(color - 40)
```
I'm so happy to see that all this stuff with bit manipulation went to separate method 👍

dshil commented on gits/terminal.py:
```diff
@@ -589,48 +648,50 @@ def write(self, s):
 
-            if i == self._cur_y * w + self._cur_x:
-                bg, fg = 1, 7
+            if i == self._cur_y * cols + self._cur_x:
```
Could you add more comments describing this if statement? Or simply extract it to the separate method with a human readable name.

eugulixes replied:
It's coloring the cursor. I'll comment the `if` statement when fix #65 since the current PR is not connected with the cursor.

dshil replied:
Ok, I got it.

dshil commented on gits/terminal.py:
```diff
@@ -589,48 +648,50 @@ def write(self, s):
+                current_classes.append('blink')
+
+            if self._screen[i] & (1 << BOLD_BIT):
+                current_classes.append('bold')
```
I'd extract
```python
            if self._screen[i] & (1 << UNDERLINE_BIT):
                current_classes.append('underline')

            if self._screen[i] & (1 << BLINK_BIT):
                current_classes.append('blink')

            if self._screen[i] & (1 << BOLD_BIT):
                current_classes.append('bold')
```
to:

```python
def _determine_cell_mode(cell):
    if cell & (1 << UNDERLINE_BIT):
         return 'underline'

    if cell & (1 << BLINK_BIT):
         return 'blink'

    if cell & (1 << BOLD_BIT):
         return 'bold'
```

usage:
```python
current_classes.append(self._determine_cell_mode(self._screen[i]))
```

dshil commented on gits/terminal.py:
```diff
@@ -589,48 +648,50 @@ def write(self, s):
             # If the characteristics of the current cell match the
             # characteristics of the previous cell, combine them into a group.
-            if bg != span_bg or fg != span_fg or i + 1 == h * w:
+            if span_classes != current_classes or i + 1 == rows * cols:
```
What for do we need the second part of this statement?
```python
or i + 1 == rows * cols:
```

eugulixes committed:
Amend `_set_bg_color` and `_set_fg_color` docstrings

Commit:
Move all stuff not related to colors outside of `_cap_set_color`

eugulixes committed:
Tests: fix typo

eugulixes committed:
Substitute underscore for underline

Different sources related to terminals use `underline` or `underscore`. For example, `man infocmp` uses `underline` and [1] uses both. I renamed `underscore` into `underline` because the latter is well-spread in our code.

[1] http://pubs.opengroup.org/onlinepubs/7908799/xcurses/terminfo.html

eugulixes committed:
Add tests for `blink`, `bold`, `smul` and `rmul` capabilities

eugulixes committed:
Switch to bright color scheme only if bold bit is set

dshil reviewed (approved):
LGTM
